command/registry.go

```diff
@@ -0,0 +1,176 @@
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/packer/hcl2template"
+	"github.com/hashicorp/packer/internal/registry"
+	"github.com/hashicorp/packer/internal/registry/env"
+	"github.com/hashicorp/packer/packer"
+
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/gocty"
+)
+
+const (
+	// Known HCP Packer Image Datasource, whose id is the SourceImageId for some build.
+	hcpImageDatasouceType string = "hcp-packer-image"
+	buildLabel            string = "build"
+)
+
+// SetupRegistry attempts to setup the HCP-related structures if HCP is enabled
+// for the command
+func SetupRegistry(cfg packer.Handler) hcl.Diagnostics {
+	switch cfg := cfg.(type) {
+	case *hcl2template.PackerConfig:
+		return setupRegistryForPackerConfig(cfg)
+	case *CoreWrapper:
+		return setupRegistryForPackerCore(cfg)
+	}
+
+	panic("NOPE.")
+}
+
+func setupRegistryForPackerConfig(pc *hcl2template.PackerConfig) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+
+	hasHCP := false
+	for _, build := range pc.Builds {
+		if build.HCPPackerRegistry != nil {
+			// Fail if multiple builds have a hcp_packer_registry
+			// block defined in them
+			if hasHCP {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Multiple " + buildLabel + " blocks",
+					Detail: fmt.Sprintf("For Packer Registry enabled builds, only one " + buildLabel +
+						" block can be defined. Please remove any additional " + buildLabel +
+						" block(s). If this " + buildLabel + " is not meant for the Packer registry please " +
+						"clear any HCP_PACKER_* environment variables."),
+					Subject: &build.HCL2Ref.DefRange,
+				})
+				return diags
+			}
+			hasHCP = true
+		}
+	}
+
+	if env.IsPAREnabled() {
+		hasHCP = true
+	}
+
+	// If we don't have HCP enabled, we can leave right away
+	if !hasHCP {
+		return diags
+	}
+
+	// Creates a bucket if either a hcp_packer_registry block is set or the HCP
+	// Packer registry is enabled via environment variable
+
+	var err error
+	pc.Bucket, err = registry.NewBucketWithIteration(registry.IterationOptions{
+		TemplateBaseDir: pc.Basedir,
+	})
+
+	if err != nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Summary:  "Unable to create a valid bucket object for HCP Packer Registry",
+			Detail:   fmt.Sprintf("%s", err),
+			Severity: hcl.DiagError,
+		})
+
+		return diags
+	}
+
+	pc.Bucket.LoadDefaultSettingsFromEnv()
+
+	for _, build := range pc.Builds {
+		build.HCPPackerRegistry.WriteToBucketConfig(pc.Bucket)
+
+		// If at this point the bucket.Slug is still empty,
+		// last try is to use the build.Name if present
+		if pc.Bucket.Slug == "" && build.Name != "" {
+			pc.Bucket.Slug = build.Name
+		}
+
+		// If the description is empty, use the one from the build block
+		if pc.Bucket.Description == "" {
+			pc.Bucket.Description = build.Description
+		}
+
+		for _, source := range build.Sources {
+			pc.Bucket.RegisterBuildForComponent(source.String())
+		}
+	}
+
+	return getHCPImageDataSources(pc)
```

Comment:
Given the name of this function I wouldn't expect it to be modifying attributes for the underlying bucket. Maybe this should be a method on the bucket that takes some data source values and inserts them into the SourceImagesToParentIterations map.

Reply:
That would be a good alternative to the current implementation too, I'll see what can be done

Reply:
For context, initally I was aiming to return a list of `hcp_packer_image.DatasourceOutput`, and defer the addition to the Bucket to the caller, but gave up when I saw the structure of the Datasource's `Values()`, hence the weird name.

Reply:
Update: removed the function entierly, it's not necessary here.